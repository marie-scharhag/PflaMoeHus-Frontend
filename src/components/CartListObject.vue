<template>
  <div id="line">
    <div class="productobject">
      <div class="picture">
        <img v-bind:src="'/api/product/picture/'+ppath.id" alt="Picture" id="pic" />
      </div>
      <div class="information">
        <ul>
          <li id="prName">{{ pname }}</li>
          <li id="prPrice">{{ pprice }} €</li>
          <li id="prNr">
            <span>Pnr: {{ particlenr }} </span>
          </li>
          <li id="inTotal">
            <span>Gesamtpreis: {{ ptotal }} €</span>
          </li>
          <select
            :value="pamount"
            name="pamount"
            id="amount"
            @change="amChange($event.target.value)"
          >
            <option
              v-for="item in pavailable"
               :value="[
               item
              ]"
              :key="item.id"
            >
            {{ item }}
             </option>
          </select>
        </ul>
      </div>
      <div class="close">
        <button id="delete" @click="trash()">
          <img src="../assets/closeImg.png" />
        </button>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import "@/service/Product";
import { defineComponent, computed } from "vue";
import { useCartStore } from "../service/CartStore";
import { useProduct } from "../service/ProductStore";

export default defineComponent({
  name: "CartListObject",
  props: {
    product: Object,
  },

  setup(props) {
    const {
      getAmount,
      changeAmount,
      deleteProduct,
    } = useCartStore();

    const { getProductByArtNr } = useProduct();
    /**
     * picture of a product
     */
    const ppath = computed(() => {
      if (props.product)
        return (getProductByArtNr(props.product[0]) as Product).allPictures[0];
    });
    /**
     * name of a product
     */
    const pname = computed(() => {
      if (props.product)
        return (getProductByArtNr(props.product[0]) as Product).name;
    });
    /**
     * price of a product
     */
    const pprice = computed(() => {
      if (props.product)
        return (getProductByArtNr(props.product[0]) as Product).price;
    });
    /**
     * articlenumber of a product
     */
    const particlenr = computed(() => {
      if (props.product)
        return (getProductByArtNr(props.product[0]) as Product).articlenr;
    });
    /**
     * number of times a product is still available
     */
    const pavailable = computed(() => {
      if (props.product)
        return (getProductByArtNr(props.product[0]) as Product).available;
    });
    /**
     * amount of times a product was placed into the cart
     */
    const pamount = computed(() => {
      if (props.product) return getAmount(props.product[0]);
    });
    /**
     * total price of the cart
     */
    const ptotal = computed(() => {
      if (props.product) {
        const t = getAmount(props.product[0]);
        if (t) {
          return (
            Math.round(
              (getProductByArtNr(props.product[0]) as Product).price *
                t *
                Math.pow(10, 2)
            ) / Math.pow(10, 2)
          );
        
      }
    }
    });
  /**
   * changes the amount of a given product in the cart
   * @param am new amount
   */
    function amChange(am: number): void {
      if (props.product) {
          changeAmount(props.product[0], Number(am));
      }
    }
    /**
     * deletes a product from the cart
     */
    function trash(): void {
      if (props.product) deleteProduct(props.product[0]);
    }

    return {
      amChange,
      trash,
      ppath,
      pname,
      pprice,
      particlenr,
      pavailable,
      pamount,
      ptotal,
    };
  },
});
</script>

<style lang = "scss" scoped>
ul{
    list-style: none;
} 
.information li{
    padding-top: 2.5%;
    padding-bottom: 2.5%;
    text-align: left;
} 
#prName{
    font-size: 16px;
    font-weight: bold;
    padding-bottom: 0%;
} 
#prPrice{
    font-size: 16px;
    padding-top: 0%;
} 
#prNr{
    font-size: 12px;
    padding-bottom: 0%;
} 
#inTotal{
    font-size: 12px;
    padding-top: 0%;
} 
#amount{
    float: left;
    font-size: 14px;
} 
.productobject{
    margin-left: 2em;
    margin-right: 2em;
    margin-top: 1.5em;
    margin-bottom: 1.5em;
    min-width: 35em;
    float: left;
} 
#delete{
    border-style: none;
    background: none;
    &:focus{
        outline: none;
    } 
} 

#pic{
    float: left;
    width: 5em;
} 
#line{
    border-bottom: 2px solid $color-grey2;
}  

.information{
    float: left;
    margin-left: 5%;
} 

.close{
    float: right;
} 
</style>