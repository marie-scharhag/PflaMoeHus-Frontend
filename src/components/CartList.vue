<template>
    <div class= "cartlist">
        <ul class="list">
         <CartListObject :product="pr" v-for="pr  in productList" :key="pr.id" />
        </ul>
        
        <div class="inTotalDiv">
            <span class="inTotal">Gesamtsumme: {{inTotal}} â‚¬</span>
        </div>
    </div>
</template>
<script lang ="ts">
import CartListObject from "../components/CartListObject.vue"
import {computed} from 'vue'; 
import {useCartStore} from '../service/CartStore'

export default {
  components: {
    CartListObject,
  },
  setup() {
    const { list, totalPrice } = useCartStore();
    /**
     * the total price of all products in cart
     */
    const inTotal = computed(() => {
      return totalPrice();
    });
    /**
     * all products in cart
     */
    const productList = computed(() => {
      return Array.from(list.value.entries());
    });

    return { inTotal, productList };
  },
};
</script>
<style lang="scss" scoped>

.list{
    display: inline-grid;
    grid-template-columns: auto;
    margin-right: 2.8rem;
}
.list:after{
    content: "";
    display: block;
}

.inTotal {
    font-size: 1.3em;
}

</style>