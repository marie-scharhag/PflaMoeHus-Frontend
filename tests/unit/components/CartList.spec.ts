import { shallowMount, VueWrapper } from "@vue/test-utils"
	
import CartList from "@/components/CartList.vue"
import CartListObject from "@/components/CartListObject.vue"
	
describe('Test CartList.vue', () => {
		
    let cmp: VueWrapper<any>

    beforeEach(() => {
        cmp = shallowMount(CartList)
    })

	
    test('has the expected html structur', () => {
        expect(cmp.element).toMatchSnapshot()
    })
    
    
    test('find out if Component CartListobject exists ', () => {
	
        const wrapper = shallowMount(CartList)
        wrapper.findComponent({ name: 'CartListObject' }) // returns a VueWrapper
        wrapper.findComponent(CartListObject) // returns a VueWrapper
	
        expect(() => wrapper.getComponent('.not-there')).toThrowError()
    })
})