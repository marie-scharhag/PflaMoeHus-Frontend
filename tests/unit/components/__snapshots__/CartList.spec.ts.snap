// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CartList.vue has the expected html structur 1`] = `
<div
  class="cartlist"
>
  <ul
    class="list"
  >
    
    
  </ul>
  <div
    class="inTotalDiv"
  >
    <span
      class="inTotal"
    >
      Gesamtsumme: 0 â‚¬
    </span>
  </div>
</div>
`;
